lib/capybara/result.rb

```diff
@@ -25,27 +25,74 @@ class Result
 
     def initialize(elements, query)
       @elements = elements
-      @result = elements.select { |node| query.matches_filters?(node) }
-      @rest = @elements - @result
+      @result_cache = []
+      @results_enum = lazy_select_elements { |node| query.matches_filters?(node) }
       @query = query
     end
 
-    def_delegators :@result, :each, :[], :at, :size, :count, :length,
-                   :first, :last, :values_at, :empty?, :inspect, :sample, :index
+    def_delegators :full_results, :size, :length, :last, :values_at, :inspect, :sample
+
+    alias :index :find_index
+
+    def each(&block)
+      @result_cache.each(&block)
+      loop do
+        next_result = @results_enum.next
+        @result_cache << next_result
+        block.call(next_result)
+      end
+      self
+    end
+
+    def [](idx)
```

Comment:
`Array#[]` supports also `[start, length] ` and `[range]` forms.
Do you think it would be better to take `*args` and call `full_results` if first item of args is not Integer >= 0 (for compatibility)?